temporal-sdk/src/main/java/io/temporal/internal/worker/LocalActivityWorker.java

```diff
@@ -180,149 +243,375 @@ public void handle(LocalActivityTask task) throws Exception {
+              ProtobufTimeUtils.toM3DurationSinceNow(activityTask.getScheduledTime());
+          metricsScope.timer(MetricsType.LOCAL_ACTIVITY_SUCCEED_E2E_LATENCY).record(e2eDuration);
+          executionContext.callback(LocalActivityResult.completed(activityHandlerResult));
+          return;
+        }
+
+        if (activityHandlerResult.getTaskCanceled() != null) {
+          executionContext.callback(LocalActivityResult.cancelled(activityHandlerResult));
+          return;
+        }
+
+        Preconditions.checkState(
+            activityHandlerResult.getTaskFailed() != null,
+            "One of taskCompleted, taskCanceled or taskFailed must be set");
+
+        Throwable executionThrowable = activityHandlerResult.getTaskFailed().getFailure();
+        Failure executionFailure =
+            activityHandlerResult.getTaskFailed().getTaskFailedRequest().getFailure();
+
+        if (isNonRetryableApplicationFailure(activityHandlerResult)) {
+          executionContext.callback(
+              failed(
+                  activityId,
+                  RetryState.RETRY_STATE_NON_RETRYABLE_FAILURE,
+                  executionFailure,
+                  currentAttempt,
+                  null));
+          return;
+        }
+
+        if (executionThrowable instanceof Error) {
+          // TODO Error inside Local Activity shouldn't be failing the local activity call.
+          //  Instead we should fail Workflow Task. Implement a special flag for that in the result.
+          //          task.callback(executionFailed(activityHandlerResult,
+          // RetryState.RETRY_STATE_MAXIMUM_ATTEMPTS_REACHED, currentAttempt));
+          // don't just swallow Error from activities, propagate it to the top
+          throw (Error) executionThrowable;
+        }
+
+        if (isRetryPolicyNotSet(activityTask)) {
+          executionContext.callback(
+              failed(
+                  activityId,
+                  RetryState.RETRY_STATE_RETRY_POLICY_NOT_SET,
+                  executionFailure,
+                  currentAttempt,
+                  null));
+          return;
+        }
+
+        RetryOptions retryOptions = RetryOptionsUtils.toRetryOptions(activityTask.getRetryPolicy());
+
+        if (RetryOptionsUtils.isNotRetryable(retryOptions, executionThrowable)) {
+          executionContext.callback(
+              failed(
+                  activityId,
+                  RetryState.RETRY_STATE_NON_RETRYABLE_FAILURE,
+                  executionFailure,
+                  currentAttempt,
+                  null));
+          return;
+        }
+
+        if (RetryOptionsUtils.areAttemptsReached(retryOptions, currentAttempt)) {
+          executionContext.callback(
+              failed(
+                  activityId,
+                  RetryState.RETRY_STATE_MAXIMUM_ATTEMPTS_REACHED,
+                  executionFailure,
+                  currentAttempt,
+                  null));
+          return;
+        }
```

Comment:
Could combine these all by just setting a retry state local var and then passing that into the failed call -- looks like the next few too

Reply:
Good eye, will refactor it.